😺 2단계 - 예외 처리 및 리팩터링 리뷰 요청

parkjun5 opened the pull request:
안녕하세요! 주말 잘 보내고 계신가요?

두 번째 단계에서 예외 처리와 리팩터링, 그리고 테마 기능 추가 구현을 완료하여 리뷰를 요청드립니다.

테이블 수정이나 도메인 수정을 할 때, 기존 테스트들이 모두 실패하여 수정이 필요하니 어려움이 있었습니다.

리뷰 잘 부탁드립니다.🙇‍♂️

boorownie committed:
fix: theme 추가로 인한 링크 추가 및 기존 링크 수정

boorownie committed:
feat: theme 추가로 인한 html 및 js 파일 생성

parkjun5 committed:
refactor: 리뷰 내용 적용

 - 패키지 명 변경 time -> reservationtime
 - delete 메서드 명 변경 delete -> deleteById
 - usingRecursiveComparison 적용

parkjun5 committed:
Merge branch 'main' into step2

parkjun5 committed:
refactor: change reservation and add Theme

parkjun5 committed:
refactor: add theme support to reservation application, repository, and UI layers

parkjun5 committed:
feat: Implement Theme Application Layer

- Add ThemeDeleter to delete theme.
- Add ThemeFinder to find themes.
- Add ThemeSaver to save new theme.

parkjun5 committed:
feat: Define Theme Domain

parkjun5 committed:
feat: Implement Theme Presentation Layer

parkjun5 committed:
test: Add more Test for Theme Domain

parkjun5 committed:
feat: Add GlobalExceptionHandler to handle application errors and define specific exceptions

parkjun5 committed:
etc: add theme page and change admin pages uri

parkjun5 committed:
test: Add theme application tests

parkjun5 committed:
polishing

xrabcde commented on src/main/java/roomescape/apply/reservation/application/ReservationRecorder.java:
```diff
@@ -5,24 +5,30 @@
+    public ReservationRecorder(ReservationRepository reservationRepository, ReservationTimeFinder reservationTimeFinder,
+                               ThemeFinder themeFinder) {
         this.reservationRepository = reservationRepository;
         this.reservationTimeFinder = reservationTimeFinder;
+        this.themeFinder = themeFinder;
     }
 
     public ReservationResponse recordReservationBy(ReservationRequest request) {
         final ReservationTime time = reservationTimeFinder.findOneById(request.timeId());
-        final Reservation reservation = Reservation.of(request.name(), request.date(), time);
+        final Theme theme = themeFinder.findOneById(request.themeId());
```
중복 예약에 대해서도 처리가 필요할 것 같아요 !

<img width="600" alt="image" src="https://github.com/next-step/spring-roomescape-reservation/assets/56033755/c5cbae4c-90ba-4805-a2f0-52614c25179e">

<img width="600" alt="image" src="https://github.com/next-step/spring-roomescape-reservation/assets/56033755/11cd94f0-7780-49a1-a30c-28fce26ca0b3">

parkjun5 replied:
추가된 정책을 확인하지 않았네요 바로 추가할께요!

xrabcde commented on src/main/resources/data.sql:
```diff
@@ -5,12 +5,23 @@ CREATE TABLE reservation_time
+    theme_id BIGINT,                                        -- 컬럼 추가
     PRIMARY KEY (id),
-    FOREIGN KEY (time_id) REFERENCES reservation_time (id) -- 외래키 추가
+    FOREIGN KEY (time_id) REFERENCES reservation_time (id),
+    FOREIGN KEY (theme_id) REFERENCES theme (id)            -- 외래키 추가
```
예약이 존재하는 테마/시간을 삭제하려고 하면 어떻게 될까요?

parkjun5 replied:
참조 무결성 유지 문제가 발생할 것 같네요!
그리고 1단계에서도 말씀해주셨는데, 처리를 하지 않았네요

이 부분 삭제하기 전에 사용하고 있는 예약이 있는지 체크하고 삭제하는 로직으로 변경하였습니다!

xrabcde commented on src/main/java/roomescape/support/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,48 @@
+
+    private static final Logger logger = LoggerFactory.getLogger(GlobalExceptionHandler.class);
+
+    @ExceptionHandler({MethodArgumentNotValidException.class, IllegalArgumentException.class})
+    public ResponseEntity<ExceptionResponse> illegalArgumentException(IllegalArgumentException e) {
+        logger.error("IllegalArgumentException caught: ", e);
+        return ResponseEntity.badRequest().body(new ExceptionResponse(400, "Bad Request"));
```
요 핸들러가 동작하던가요? 예외 케이스에 대해서도 테스트코드를 추가해주시면 좋을 것 같습니다 !! 

<img width="431" alt="image" src="https://github.com/next-step/spring-roomescape-reservation/assets/56033755/6c57c019-acaf-455c-bf55-f54449475d88">

<img width="1657" alt="image" src="https://github.com/next-step/spring-roomescape-reservation/assets/56033755/c20a6474-5b7b-45db-8493-dfe26445fde6">

parkjun5 replied:
핸들러가 작동안하는 줄 알고 계속 찾고 있었는데.. 놀랍게도 작동은 하지만 제가 예외 처리에서 스택트레이스를 전부 출력해서 안한 듯 한 상태로 보였네요.

500 에러가 출력되는 이유는 HttpMessageNotReadableException처리를 하지 않아 RuntimeExcpetion으로 설정되어 버리기 떄문인 듯 합니다. 테스트 추가하겠습니다.

xrabcde commented on src/main/java/roomescape/apply/reservation/application/excpetion/NotFoundReservationException.java:
```diff
@@ -0,0 +1,7 @@
+
+import java.util.NoSuchElementException;
+
+public class NotFoundReservationException extends NoSuchElementException {
+
```
커스텀 예외 정의 👍👍

xrabcde commented on src/main/java/roomescape/apply/reservation/ui/ReservationController.java:
```diff
@@ -26,7 +26,7 @@ public ReservationController(ReservationRecorder reservationRecorder, Reservatio
 
     @GetMapping
-    public ResponseEntity<List<ReservationResponse>> getReservation() {
+    public ResponseEntity<List<ReservationResponse>> getReservations() {
```
👍👍

xrabcde commented on src/main/java/roomescape/apply/reservation/ui/dto/ReservationResponse.java:
```diff
@@ -1,17 +1,20 @@
         ReservationTimeResponse time
 ) {
-    public static ReservationResponse from(Reservation reservation, ReservationTime reservationTime) {
+    public static ReservationResponse from(Reservation reservation, Theme theme, ReservationTime reservationTime) {
```
Reservation 안에도 theme과 reservationTime이 필드로 있는데 각각 따로 받으시는 이유가 있나요? 
(그냥 궁금해서 여쭤봅니다... )

parkjun5 replied:
냅냅

제 생각에 물론 Reservation 안에 필드로 전부 가지고는 있지만, DTO 내부 로직에서 해당 필드를 가져오는 로직을 사용하는 것보다는 파라미터로 넘겨주는게 더 가독성 있고 테스트하기 편하다고 생각합니다.
필요하지 않는 경우에는 null 혹은 Mock객체만 넣어주면 사용하기 편하더라구요.

그리고 막상 간단한 코드라도 Presentation Layer에 넣다보면 나중에는 값을 변환해주는 stream 같은 로직들이 DTO에 들어가는 경우도 종종 발생하더라구요..(빠른 개발을 위해?)

그래서 애초부터 파리미터로 받도록 하고 DTO를 생성하는 쪽에서 파라미터로 넣어주는 것이 유지보수하기에도 더 좋을 것 같다는 막연한 생각도 있구요..

리뷰어님께서 보통 이런 경우에 어떻게 처리하시는지에 대해서도 궁금합니다. 더 나은 방법이 있다면 배우고 싶습니다.

xrabcde replied:
오호 네 다양한 관점에서 생각을 말씀해주셔서 감사합니다 (_ _)
저는 ReservationResponse 의 역할은 `Reservation 이 갖고 있는 정보들을 응답해준다` 
라고 생각해서 reservation 의 필드들을 활용할 것 같다고 생각했는데 
세준님 말씀도 들어보니 공감이 되네요 ㅎㅎ 취향의 영역인 것 같습니다 :)

xrabcde reviewed (changes requested):
안녕하세요 세준님 ~ 2단계도 빠르게 구현해주셨네요 ㅎㅎ 
예외처리 관련해서 몇 가지 보완해보면 좋을 부분이 있어 코멘트 드렸습니다 
확인해보시고 다시 리뷰요청 보내주세요 ~ 🙇🏻‍♀️

parkjun5 committed:
refactor: 리뷰 내용 적용

 - 예약 시간, 테마 삭제시 예약이 존재하는 지 확인 후 삭제하도록 변경(예약이 있으면 삭제 불가하도록 Table 변경)
 - GlobalExceptionHandler가 좀더 명확하게 처리하도록 변경
    - 해당 변경점을 위한 테스트 추가

xrabcde commented on src/main/java/roomescape/apply/reservation/application/ReservationRecorder.java:
```diff
@@ -16,19 +17,27 @@ public class ReservationRecorder {
     }
 
     public ReservationResponse recordReservationBy(ReservationRequest request) {
+        boolean alreadyExisted = reservationFinder.doesReservationExist(request.timeId(), request.themeId());
+        if (alreadyExisted) {
+            throw new DuplicateReservationException();
+        }
```
중복 예약을 검증하는 로직은 별도 메서드로 분리해줘도 좋겠네용

parkjun5 replied:
별도로 메서드를 나눈다면 private 메서드가 만들어지니, 
DuplicationChecker 클래스를 만들고 중복 체크 관리를 책임져주면 좋을 것 같아요

xrabcde commented on src/main/java/roomescape/apply/theme/application/ThemeDeleter.java:
```diff
@@ -1,20 +1,31 @@
 
     public void deleteThemeBy(long id) {
         final long existId = themeRepository.checkIdExists(id).orElseThrow(NotFoundThemeException::new);
+        boolean isReferenced = reservationFinder.findAnyByThemeId(id).isPresent();
+        if (isReferenced) {
+            throw new ThemeReferencedException();
+        }
```
여기두용

xrabcde commented on src/main/java/roomescape/apply/theme/application/exception/ThemeReferencedException.java:
```diff
@@ -0,0 +1,10 @@
+
+public class ThemeReferencedException extends IllegalArgumentException {
+
+    public static final String DEFAULT_MESSAGE = "테마가 다른 예약에서 참조되고 있어 삭제할 수 없습니다.";
+
```
이 메시지를 사용자가 한 번에 이해할 수 있을까요?
좀 더 직관적인 문구가 있을 것 같아요 ☺️

parkjun5 replied:
사용자에게 직관적인 메시지를 작성하는 건 정말 어렵네요. 🤔
`예약이 존재하는 테마입니다.` 가 좀 더 직관적인 것 같아요!

xrabcde commented on src/test/java/roomescape/apply/reservation/application/ReservationRecorderTest.java:
```diff
@@ -63,4 +64,18 @@ void recordReservationBy() {
+        Theme theme = themeRepository.save(theme());
+        ReservationRequest request = reservationRequest(time.getId(), theme.getId());
+        // when && then
+        Assertions.assertDoesNotThrow(() -> reservationRecorder.recordReservationBy(request));
+        assertThatThrownBy(() -> reservationRecorder.recordReservationBy(request))
+                .isInstanceOf(DuplicateReservationException.class)
+                .hasMessage(DuplicateReservationException.DEFAULT_MESSAGE);
```
assertThatThrownBy를 활용한 예외 케이스 검증 👍👍 

두 가지 케이스를 한 테스트에서 모두 확인해주고 있는데 
꼭 필요한 경우가 아니라면 한 테스트에서는 한 케이스만 확인해주는 게 좋을 것 같습니다 !

parkjun5 replied:
assertDoesNotThrow로 이전에는 작동했다가, 이후에는 에러가 Throw 한다는 것을 표현하려고 했는데, 오히려 두개의 검증을 하는 테스트가 되었네요. 
`Assertions.assertDoesNotThrow(() -> reservationRecorder.recordReservationBy(request));`부분은 현재 테스트에서 검증하고 싶은 부분이 아니니 when 절에 `reservationRecorder.recordReservationBy(request);`만 포함하면 되겠네요.

xrabcde reviewed (approved):
안녕하세요 세준님 ~ 피드백 내용을 잘 반영해주셨네요 ㅎㅎ 
첫 주차 미션 진행하시느라 고생많으셨습니다 
몇 가지 코멘트 남겨드렸는데 시간 되실 때 확인해주세요 😉
이어서 다음 미션들도 화이팅입니다 ~!
